// @sap/html5-repo-mock/lib/services/portal/inbounds.js
module.exports = tiles => {
    const inbounds = {};
    tiles.forEach(tile => {
        if(tile.properties && tile.properties.targetURL && tile.properties.targetURL.startsWith('#')) {
            const idx = tile.properties.chipId.lastIndexOf('_');
            const componentId = tile.properties.chipId.slice('catalogTile_'.length, idx);
            const [semanticObject, action] = tile.properties.targetURL.slice(1).split('-');
            inbounds[semanticObject.toLowerCase() + action[0].toUpperCase() + action.slice(1).toLowerCase()] = {
                semanticObject,
                action,
                resolutionResult: {
                    additionalInformation: 'SAPUI5.Component=' + componentId,
                    applicationType: 'SAPUI5',
                    url: `/${tile.id.replace(/_([^_]+)$/, '-$1')}`
                },
                signature: {
                    parameters: {},
                    additionalParameters: 'allowed'
                },
                title: tile.properties.title
            };
        }
    });
    return inbounds;
};
